### PR TITLE
function test harness option to skip tests locally

### PR DESCRIPTION
**Pull request by natasha41575**
Add option to function test config: `skipIfLocal`. If this is set, then the test runner will check for an env variable `IN_CI`. If `IN_CI` is not set to true, then we assume that we are running the test locally and skip the test.

This also sets `IN_CI=true` to the CI job.

### CODE REVIEW

**Comment by natasha41575** [comment]
cc @mengqiy

File: pkg/test/runner/README.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ in `.expected`:
     used to run the test. Default: `render`.
   - `exitCode`: The expected exit code for the command. Default: 0.
   - `skip`: Runner will skip the test if `skip` is set to true. Default: false.
+  - `skipIfLocal`: Runner will skip the test if the env var IN_CI is "false" or not set. Default: false.
```
Comment (mengqiy) [inline pkg/test/runner/README.md:25]: I'd suggest to make the 2 names more consistent in terms of naming. For example, `skipIfNotInCI` with `KPT_E2E_IN_CI`.

File: .github/workflows/go.yml [hunk 2]
```diff
@@ -46,7 +46,7 @@ jobs:
         run: |
           git config --global user.email you@example.com
           git config --global user.name Your Name
-          make all
-          make test-docker
+          IN_CI=true make all
```
Comment (mengqiy) [inline .github/workflows/go.yml:49]: Does `make all` need it?

File: .github/workflows/go.yml [hunk 3]
```diff
@@ -46,7 +46,7 @@ jobs:
...
+          IN_CI=true make test-docker
```
Comment (mengqiy) [inline .github/workflows/go.yml:50]: Do we have any tests in kpt repo relying on this?

File: .github/workflows/go.yml (hunk 2 above)
Comment (natasha41575) [reply .github/workflows/go.yml:49]: I think you're right that `make all` doesn't need it, will remove

File: .github/workflows/go.yml (hunk 3 above)
Comment (natasha41575) [reply .github/workflows/go.yml:50]: There are no tests in the kpt repo that use this; I did some manual testing. For further testing maybe either you could pull this branch to do your own manual testing or I could commit a dummy test just to exercise the new option?

File: pkg/test/runner/README.md (hunk 1 above)
Comment (natasha41575) [reply pkg/test/runner/README.md]: done!

**Comment by natasha41575** [comment]
@mengqiy PR is updated, PTAL

**Comment by natasha41575** [comment]
cc @yuwenma

File: pkg/test/runner/config.go [hunk 4]
```diff
@@ -85,6 +85,9 @@ type TestCaseConfig struct {
 	// Skip means should this test case be skipped. Default: false
 	Skip bool `json:"skip,omitempty" yaml:"skip,omitempty"`
 
+	// SkipIfNotInCI means should this test case be skipped if the env var KPT_E2E_IN_CI is "false" or not set. Default: false
```
Comment (yuwenma) [inline pkg/test/runner/config.go:88]: "means this test case should be..."

**Comment by yuwenma** [comment]
What is the scenarios that requires this condition? The solution seems to be a temporary workaround.

**Comment by natasha41575** [comment]
> What is the scenarios that requires this condition? The solution seems to be a temporary workaround.

This is needed for the function catalog render-helm-chart private repo tests. The credentials are populated by Github secrets stored in CI, so it won't work when running locally.

File: pkg/test/runner/config.go (hunk 4 above)
Comment (natasha41575) [reply pkg/test/runner/config.go]: Thanks for the catch! Copy paste mistake. Fixed.

**Comment by yuwenma** [comment]
> > What is the scenarios that requires this condition? The solution seems to be a temporary workaround.
> 
> This is needed for the function catalog render-helm-chart private repo tests. The credentials are populated by Github secrets stored in CI, so it won't work when running locally.

My main concern is that the SKIP and SKIPIfNotInCI config have overlapping use cases so it makes the code hard to maintain and extend. 
 
From the render-helm-charts test, do you think we can be simplified it as:
 ```go
func TestYourRenderHelmCharts():
      if testutil.IsDevLocal():
            pass

func IsDevLocal():
      _, githubCI := os.LookupEnv("CI")
      return !githubCI
```
[CI](https://docs.github.com/en/actions/learn-github-actions/environment-variables) is the GitHub default env var and is always set to true. So no manual change is needed and we do not complicate the runner skip config. WDYT?

**Comment by natasha41575** [comment]
> CI is the GitHub default env var and is always set to true. So no manual change is needed and we do not complicate the runner skip config. WDYT?

I didn't know GH had a default CI var! That's very useful. That sounds good to me, I will update it.

**Comment by yuwenma** [comment]
I see. Then that's fair to me. Can we change `skipIfNotInCI` to something like `SkipLocal`?  "skip" is like a NOT condition, and we can always improve code readability by avoiding double negation. :)

**Comment by natasha41575** [comment]
> I see. Then that's fair to me. Can we change skipIfNotInCI to something like SkipLocal? "skip" is like a NOT condition, and we can always improve code readability by avoiding double negation. :)

Done!

**Comment by natasha41575** [comment]
@yuwenma based on our chat offline today, I don't think we need this anymore. Closing.

**Pull request closed without merging**